### ISSUE TITLE
关于prompt优化过程的疑问

### ISSUE THREAD

**Issue by zenithc-git** [MorningStarOvO/ProAPO#5]
您好，感谢您开源本项目，这对于使用LLM辅助clip进行分类的问题上具有巨大的准确性以及时间效率的提升与贡献

但是我在阅读代码的时候发现，虽然在优化过程中（对库中预设prompt自动add或者删或者拼接的过程）是只用了fewshot训练集来打分的，但是中间插了多次用test评估然后选最优，每个inner epoch结束后都会进行一次test数据集上的评估，如果出现了更高的TOP1 acc便会记录结果和该轮次使用的prompt。这样的话也就意味着其实在test数据集上得分最高的一次结果不一定是整个流程最后一次循环优化出的prompt在test上评估得到的结果对吗？最优的prompt也不一定是最后一个epoch得到的prompt是吗？

如果是这样的话，是否算使用测试集来辅助训练造成了测试集的泄露呢？

最后还是再次感谢您们做的工作，我对贵项目和代码的理解也有许多不全面之处，希望可以得到您的回复

**Comment by MorningStarOvO** [comment]
您好，

非常感谢您对我们工作的认可以及如此细心的观察和深入的代码阅读！您提出的问题非常有价值。

关于您的疑问，我需要说明以下几点：

首先，您的观察是正确的。 目前代码输出的 exp best 结果确实会带来误解。这里输出的是当前所有 epoch 中最优的结果，但实际上论文里报告的是最后一个 epoch 的结果。我们已经在 class_specific_selection_with_same_template.py 的第 297 行中进行了修复，现在代码会正确输出最后一个 epoch 的结果。

关于每个 epoch 都进行测试集评估的设计初衷：
1. 保存 "detailed epoch" 的结果，记录随着 epoch 的变化我们方法的性能变化曲线，这有助于分析方法的收敛特性
2. 方便给出最终的 Upper Bound 结果，作为性能分析的参考
3. 观察迭代过程是否存在严重的过拟合现象，及时发现训练中的问题

需要明确的是：这些中间的测试集评估结果仅用于分析和记录，并不参与 prompt 的选择过程。Prompt 的优化只依赖于 few-shot 训练集的结果。

再次感谢您细心的观察和宝贵的反馈！如果当前代码给您带来了不便或误解，我们深表歉意。如有任何其他问题，欢迎随时交流讨论。

Best regards

**Comment by zenithc-git** [comment]
好的，感谢您的回答，by the way，我这边还发现一些小问题，CUB数据集的split文件好像出现了一些错误，将train和test弄反了，我看了一下CUB数据集官方的readme写到“a value of 1 or 0 for <is_training_image> denotes that the file is in the training or test set, respectively.”也就是1代表测试集而0代表训练集。
但比如split文件中train的最后一行
“CUB_200_2011/images/200.Common_Yellowthroat/Common_Yellowthroat_0008_190703.jpg": "Common Yellowthroat"”，在CUB数据集的images.txt文件中对应id 11786，这个id通过查询他们的train_test_split.txt对应的是数字0，应当是测试集而非训练集

不过我没有一一检查每一行，但似乎是略有不同的

以及比如food我看到您们使用的split.json与split_zhou_Food101.json在train和test划分上也是不同的，我将代码Test流程中使用的图片路径保存并打印后发现使用的是split文件中划分的train，test而不是split_zhou_Food101.json中的这种划分，可以请问这是unintentional mistake 还是专门进行了这样的设定呢

再次感谢您的回答~

**Comment by MorningStarOvO** [comment]
您好，

非常感谢您仔细检查和反馈这些细节！🙏 

关于 CUB 数据集：我们并不是使用原始数据集自带的 train_test_split.txt，而是采用了论文 “Zero-shot learning: A comprehensive evaluation of the good, the bad and the ugly” 中提供的 Proposed Split。该划分是专门为 Zero-Shot Learning (ZSL) 任务重新定义的，具体的下载地址为：
http://datasets.d2.mpi-inf.mpg.de/xian/xlsa17.zip，
路径在 xlsa17/data/CUB/att_splits.mat 中的 split 信息。它专门针对 ZSL 任务重新划分了下 Split

关于 Food101 数据集，我们使用的是 data/dataset/food101/meta/test.txt 和 data_description/Food101/meta/train.txt 中定义的 Split，其是原始数据集所提供的，我已经上传到代码仓库中。我记得当时是看到 WaffleCLIP 或 DCLIP 使用了这一划分方式，所以用了这个。其他数据集的划分则基本延续了 CoOp 代码仓库中的配置。

再次感谢您的关注与反馈！ 🙏

**Comment by zenithc-git** [comment]
好的好的，非常感谢您的回答
